c/datasetjson.c

```diff
@@ -1630,6 +1640,395 @@ void respondWithDataset(HttpResponse* response, char* absolutePath, int jsonMode
+               
+          /* fill the table slot */
+          sem_table_entry[i].sem_ID = semaphoreID;
+          memcpy(sem_table_entry[i].mem, dsn_member.membername, 8);   /* copy in member */
+          memcpy(sem_table_entry[i].dsn, dsn_member.dsn, 44); /* copy in dsn */
+
+          /*******/
+          /* set */
+          /*******/
+          // semctl() changes permissions and other characteristics of a semaphore set
+          union semun {
+              int val;
+              struct semid_ds *buf;
+              unsigned short *array;
+          } arg;
+          arg.val = 1;
+          int     semnum = 0;    /* array index zero */
+          int semaphoreRetcode;
+
+          zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+                  "ZSS1916I %s Set semaphore %d, PID %d\n", __FUNCTION__, semaphoreID, pid);
+          semaphoreRetcode = semctl(semaphoreID, semnum, SETVAL, arg); /* set the value of our semaphore */
+          if (semaphoreRetcode == -1 )
+            {
+              printf("Failed to set semaphore\n"); 
+              zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+                "ZSS2015I %s Failed to set semaphore %d\n", __FUNCTION__, semaphoreID);
+              respondWithError(response, HTTP_STATUS_INTERNAL_SERVER_ERROR, "Failed to set semaphore");
+              return;                  
+            }
+          else
+            {
+              zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+                "ZSS2020I %s semaphore %d was set\n", __FUNCTION__, semaphoreID);
+            }
+
+          /* the HTTP GET that consumes this message insists that the response body be JSON */
+          respondWithMessage(response, HTTP_STATUS_OK,
+                     "{\"records\":[\"Enqueue dataset successful\"]}"
+                     ); 
+
+          /********/
+          /* wait */
+          /********/
+          /* WAIT! */
+          zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+                  "ZSS1916I %s wait on semaphore %d, PID %d\n", __FUNCTION__, semaphoreID, pid);
+
+          /* define semaphore operation to be performed */
+          struct sembuf semaphoreBuffer[1]; /* just one semaphore in the array */
+          struct sembuf *semaphoreOps = &semaphoreBuffer[0];
+          semaphoreBuffer[0].sem_num = 0; /* index of first and only semaphore */
+          semaphoreBuffer[0].sem_op  = 0; /* 0 = wait */
+          semaphoreBuffer[0].sem_flg = 0; /* 0 = sychronous + don't undo */
+
+          /* wait for semaphore to be posted */
+          
+          semaphoreRetcode = semop(semaphoreID, semaphoreOps, 1); /* 1 = one entry */
+                /* we are now waiting for our semaphore to be posted ... */
+
+          
+          if (semaphoreRetcode != 0 )
+          {
+              zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+                "ZSS1938I %s Failed to wait on semaphore, retcode = %d\n", __FUNCTION__, semaphoreRetcode);
```

Comment:
Should this entire logic be moved before the "respondWithMessage? If the semaphore operation is not posted then could this cause unexpected conditions?

Reply:
I think the code is right as it stands.  The test `if (semaphoreRetcode != 0 )` asks "did the semop call fail?" so it would fail to start waiting.  Otherwise it waits until it's posted.